Environment: add subsubsection on override methods, revise subsection on Monticello packages (#146)
hpi-swa-lab/SqueakByExample-english: Environment/Environment.tex

\end{code}

\noindent
These methods belong in the same package as the classes in the categories \ct{SqueakLink-Connections} and \ct{SqueakLink-Model}.
But clearly the whole of class \ct{Object} does not belong in your package!
So you need a way of putting certain \emph{methods} in a package, even though the rest of the class is in another package.
\index{package!extension}
\seeindex{extension method}{package, extension}

This is where the concept of \ind{extension methods} comes in.

> You can mark these converter methods as extension methods by placing them in a protocol (of \ct{Object}, \ct{String}, \ct{Date}, and so on) named \prot{*SqueakLink-Model-converting} (note the initial asterisk).

Reviewer: Maybe add a sentence in the direction of: "As a result, the three methods are now being managed as part of the SqueakLink package".

Author: ```suggestion
You can mark these converter methods as extension methods by placing them in a protocol (of \ct{Object}, \ct{String}, \ct{Date}, and so on) named \prot{*SqueakLink-Model-converting} (note the initial asterisk).
As a result, the three methods are now being managed as part of the SqueakLink package.
```

> % In general, any method in any protocol that starts with an asterisk, followed by a package name, followed by a hyphen, is an extension method to that package.

Reviewer: I like this repetition of the more general rule. :)

Author: ```suggestion
In general, any method in any protocol that starts with an asterisk, followed by a package name, followed by a hyphen, is an extension method to that package.
```

\subsubsection{Override methods}

Sometimes, you may also want to alter a method from another package that already exists and store the changed method in your package.
For instance, this can be required if you need to extend an existing feature or fix a bug in the standard Squeak image.

> Let's say that you want to add a menu item to an inspector's context menu for opening its SQL version, so you have edited \ct{Inspector>>addObjectItemsTo:} and now would like to check it into your package repository.

Reviewer: "opening its SQL version, so you have edited"
->
"opening its SQL version. Therefore, you have"

Author: ```suggestion
Let's say that you want to add a menu item to an inspector's context menu for opening its SQL version.
 Therefore, you have edited \ct{Inspector>>addObjectItemsTo:} and now would like to check it into your package repository.
```

\index{package!extension!overriding}
\seeindex{override method}{package, extension, overriding}

While you could place it in a simple extension protocol as above, this would introduce some problems:

> First, the method would be displayed as missing in the original package, so the package would unintentionally become dirty.

Reviewer: "become dirty" -> would be treated as modified. # dirty is somewhat tool-specific here

Author: ```suggestion
First, the method would be displayed as missing in the original package, so the package would unintentionally become \emph{dirty} (\ie it would be considered containing changes that were not yet saved to any repository).
```

Second, if you unloaded your own package again, the original method would be lost.

A better way to do this is to mark the method as an \ind{overriding method} by placing it in a protocol named \prot{*SqueakLink-menu-override}.
In general, an override protocol is an extension protocol that ends with the suffix \prot{-override}.
Usually, it's a good idea to repeat the name of the original protocol in the override protocol name.

However, beware that every change to an existing package might break other packages that depend on it.
Thus, use override methods as seldom as possible.

> Often, there are better ways to extend the functionality of a package, such as subclassing or using explicit hooks.\footnote{Actually, there is a better way to extend the menu in our example: Place a new (non-override) extension method named \ct{squeakLinkMenu:} in the \ct{Inspector} class and add the pragma \ct{<fieldListMenu>} to it.}

Reviewer: "Actually, there is an even better way to extend the menu in our example:"

Author: ```suggestion
Often, there are better ways to extend the functionality of a package, such as subclassing or using explicit hooks.\footnote{Actually, there is an even better way to extend the menu in our example: Place a new (non-override) extension method named \ct{squeakLinkMenu:} in the \ct{Inspector} class and add the pragma \ct{<fieldListMenu>} to it.}
```

\subsubsection{Surveying the contents of a package}

The combination of the \scat{SqueakLink-...} categories and the \prot{*SqueakLink-...} protocols form a package named \ct{SqueakLink}.
To be precise, the rules for what goes in a package are as follows.

A package named \ct{Foo} contains:

\begin{enumerate}\label{sec:packageRules}
	\item{} all class definitions of classes in the category \scat{Foo}, or in categories with names starting with \scat{Foo-} (or \scat{foo-}, \scat{fOO}, autc., as Monticello actually always ignores the case of the letters),
	and
	\item{} \label{env:extensions} all method definitions in any class in a protocol named \prot{*Foo} or whose name starts with \prot{*Foo-},
	and
	\item{} all methods in classes in the category \scat{Foo} or in a category whose name starts with \scat{Foo-}, \emph{except} for those methods in protocols whose names start with \prot{*}.
\end{enumerate}
\noindent
A consequence of these rules is that each class definition and each method belongs to exactly one package.

> % ct: This is NOT true if you define multiple packages that overlap each other, e.g., Kernel and Kernel-Exceptions! Should we mention that?

Reviewer: That's a tough one. It would require rewritng this whole section on the basis that packages are an implicit concept that solely exists through the naming scheme. 
In practice, for beginners this does not make a large difference. Also, you could argue that Kernel is a package including the Kernel-* packages, so the method or class does actually belong to exactly one "most-specific" package, and potentially multiple package-packages. I vote for not discussing this.

Author: ```suggestion
% ct: In fact, e.g., Kernel and Kernel-Exceptions and could be defined as overlapping packages. However, this is commonly not an intended usage of packages so we omit it for sake of simplicity here.
```

For more sophisticated versioning of packages, we use Monticello.
\index{file!filing out}

\begin{ExecuteSmalltalkScript}
	SBEScreenshotRecorder writeTo: './figures/browsePackageExtensions.png' building: [:helper | | window |
	Browser open.
	window := helper foregroundWindow.
	helper
		selectAnyListIn: window at: 'Morphic-Kernel' andScrollIntoViewBy: 1;
		selectAnyListIn: window at: 'Morph' andScrollIntoViewBy: 1.
	helper
		click: window at: (window pointAtFraction: 1 / 4 * 2 / 3 @ 0) + (7 px @ 40 px) buttons: #(right);
		mouseMoveTo: (helper menuItemMorphLabeled: 'browse package*' in: helper world) center buttons: #() in: helper world;
		mouseMoveTo: (helper menuItemMorphLabeled: 'extensions' in: helper world) center buttons: #() in: helper world].
\end{ExecuteSmalltalkScript}
\begin{figure}

> 	\includegraphics[width=\textwidth]{figures/browsePackageExtensions.png}

Author: ```suggestion
	\includegraphics[width=\textwidth]{browsePackageExtensions.png}
```
